### PR TITLE
Bug 2035757: cluster-bootstrap/alibaba: set tear-down-delay to wait kube-apiserver rolls out on AlibabaCloud

### PR DESCRIPTION
**Pull request by mtulio**
Setting up `--tear-down-delay` flag to `cluster-bootstrap` to wait `10m` until it tear down, to give time to wait for two kube-apiserver pods is available to finish the bootstrap in AlibabaCloud (only). The default value will be set to `0`.

This is caused due to a premature ending of bootkube of `cluster-bootstrap`/`kube-apiserver`  when scheduled the kube-apiserver to only one master, and the "SLB limitation where cannot be accessed by backend servers"[1]. 
[1] https://www.alibabacloud.com/help/en/doc-detail/55206.htm

This is an investigation of the bug https://bugzilla.redhat.com/show_bug.cgi?id=2035757

### CODE REVIEW

**Comment by staebler** [comment]
Here is my analysis of the steps that cause the bug.
1. One of the masters creates a kube-apiserver pod.
1. The cluster-bootstrap on the bootstrap node sees that all of the kube-apiserver pods are ready (even though there has only been 1 created so far).
1. The control plane on the bootstrap is torn down.
1. The node for the master with the kube-apiserver stops reporting heartbeat since it cannot access the api server via api-int.
1. The kube-apiserver pod stops behaving since its token is revoked.
1. Nobody can access the api server any more.

The proposed solution presented in this PR is to delay the time between steps (2) and (3) so that the kube-apiservers on the other nodes have a chance to start while the temporary control plane on the bootstrap node is still running.

This is not a long-term solution as this does not address any issues that may arise during upgrades as new apiserver revisions are rolled out. If there is a time when there is only one kube-apiserver pod running, then the cluster will get stuck with a non-responding apiserver.

**Comment by mtulio** [comment]
@staebler I just changed to use the bootstrap template. ptal?

Tests:
for `alibabacloud` :
```
$ jq -r '.storage.files[] | select(.path=="/usr/local/bin/bootkube.sh") | .contents.source' .local/clusters/tmp-alibaba/bootstrap.ign  |sed 's/data\:text\/plain\;charset\=utf\-8;base64,//g' |base64 -d |grep -A 10 CLUSTER_BOOTSTRAP_IMAGE -m2
[...]
--
        "${CLUSTER_BOOTSTRAP_IMAGE}" \
        start \
          --tear-down-early=false \
          --tear-down-delay="10m" \
          --asset-dir=/assets \
          --required-pods="${REQUIRED_PODS}"

```

for others (`aws`):
```
$ jq -r '.storage.files[] | select(.path=="/usr/local/bin/bootkube.sh") | .contents.source' .local/clusters/tmp-aws/bootstrap.ign  |sed 's/data\:text\/plain\;charset\=utf\-8;base64,//g' |base64 -d |grep -A 10 CLUSTER_BOOTSTRAP_IMAGE -m2
CLUSTER_BOOTSTRAP_IMAGE=$(image_for cluster-bootstrap)
[...]
        "${CLUSTER_BOOTSTRAP_IMAGE}" \
        start \
          --tear-down-early=false \
          --tear-down-delay="0" \
          --asset-dir=/assets \
          --required-pods="${REQUIRED_PODS}"
}

```

File: pkg/asset/ignition/bootstrap/common.go [hunk 1]
```diff
@@ -278,6 +280,18 @@ func (a *Common) getTemplateData(dependencies asset.Parents, bootstrapInPlace bo
...
+	// Set Tear-down delay for platform specific
+	var tearDownDelay string
+	switch installConfig.Config.Platform.Name() {
+	case alibabacloud.Name:
+		// tear-down set to let kube-apiserver rolls out and avoid loopback CLB limitation
+		// BZ https://bugzilla.redhat.com/show_bug.cgi?id=2035757
+		tearDownDelay = "10m"
+	default:
+		tearDownDelay = "0"
+	}
```
Comment (staebler) [inline pkg/asset/ignition/bootstrap/common.go:293]: Let's make the happy path where we don't have to set the delay more prominent.
```suggestion
	tearDownDelay := "0"
	if installConfig.Config.Platform.Name() == alibabacloud.Name {
		// tear-down set to let kube-apiserver rolls out and avoid loopback CLB limitation
		// BZ https://bugzilla.redhat.com/show_bug.cgi?id=2035757
		tearDownDelay = "10m"
	}
```

**Review by staebler — COMMENTED** [review]
Looks good. Just a nit about code structure.

**Comment by mtulio** [comment]
As we have no CI for this component yet, sharing the result of the cluster running with this version:
```bash
DEBUG Time elapsed per stage:                      
DEBUG            cluster: 2m35s                    
DEBUG          bootstrap: 1m17s                    
DEBUG Bootstrap Complete: 21m47s                   
DEBUG                API: 2m31s                    
DEBUG  Cluster Operators: 20m34s                   
INFO Time elapsed: 46m22s

NAME                                   STATUS   ROLES    AGE   VERSION
mrbkas-f7vb6-master-0                  Ready    master   56m   v1.23.0+60f5a1c
mrbkas-f7vb6-master-1                  Ready    master   54m   v1.23.0+60f5a1c
mrbkas-f7vb6-master-2                  Ready    master   55m   v1.23.0+60f5a1c

kube-apiserver-mrbkas-f7vb6-master-0         5/5     Running     0          20m
kube-apiserver-mrbkas-f7vb6-master-1         5/5     Running     0          12m
kube-apiserver-mrbkas-f7vb6-master-2         5/5     Running     0          14m
```

cc @kwoodson regarding CI PR:
https://github.com/openshift/release/pull/20841

**Review by staebler — COMMENTED** [review]
/lgtm
/approve
/label tide/merge-method-squash

**Comment by staebler** [comment]
/override ci/prow/e2e-aws-upgrade
The install is successful. The changes in this PR only affect the installation process.

**Comment by mtulio** [comment]
@staebler Not sure if the `/override` is valid for the PR lifecycle. Do we need to send `/override ci/prow/e2e-aws-upgrade` again? The last summary shows that it ran. :(

**Comment by mtulio** [comment]
/override ci/prow/e2e-aws-upgrade
Already skipped and not processed by ci bot
> /override ci/prow/e2e-aws-upgrade The install is successful. The changes in this PR only affect the installation process.

**Comment by mtulio** [comment]
> @mtulio: mtulio unauthorized: /override is restricted to Repo administrators, approvers in top level OWNERS file.

@staebler @patrickdillon please skip it again.

**Pull request merged**